src/Microsoft.AspNet.OData.Shared/Builder/FunctionConfiguration.cs

```diff
@@ -202,6 +202,47 @@ public FunctionConfiguration SetBindingParameter(string name, IEdmTypeConfigurat
             return this;
         }
 
+        /// <summary>
+        /// Adds subtypes to the list of derived type constraints for the return type.
+        /// </summary>
+        /// <param name="subtypes">The subtypes for which the constraint needs to be added.</param>
+        /// <returns>Updated configuration object.</returns>
+        public FunctionConfiguration HasDerivedTypeConstraintsForReturnType(params Type[] subtypes)
+        {
+            if (ReturnType == null)
+            {
+               throw Error.InvalidOperation(SRResources.ReturnTypeOfOperationNotSpecified);
+            }
+
+            if (ReturnTypeConstraints == null)
+            {
+                ReturnTypeConstraints = new DerivedTypeConstraintConfiguration(ReturnType.ClrType);
+            }
+
+            ReturnTypeConstraints.AddConstraints(subtypes);
+            return this;
+        }
+
+        /// <summary>
+        /// Adds TDerived to the list of derived type constraints for the return type. 
+        /// </summary>
+        /// <returns>Updated configuration object.</returns>
+        public FunctionConfiguration HasDerivedTypeConstraintForReturnType<TDerivedType>()
+        {
+            if (ReturnType == null)
+            {
+                throw Error.InvalidOperation(SRResources.ReturnTypeOfOperationNotSpecified);
+            }
+
+            if (ReturnTypeConstraints == null)
+            {
+                ReturnTypeConstraints = new DerivedTypeConstraintConfiguration(ReturnType.ClrType);
+            }
+
+            ReturnTypeConstraints.AddConstraint<TDerivedType>();
```

Comment:
why don't you move all these codes into base class OperationConfiguration?

Reply:
To avoid the cast while using the fluent APIs. I can have an internal implementation on operation configuration but it is only 3 lines of code. 

---
In reply to: [401813407](https://github.com/OData/WebApi/pull/1977#discussion_r401813407) [](ancestors = 401813407)